app/src/main/java/me/saket/dank/ui/UrlRouter.java

```diff
@@ -126,10 +126,20 @@ public Intent intent(Context context) {
 
           // Smart linking.
           if (defaultWebBrowser == DefaultWebBrowser.DANK_INTERNAL_BROWSER) {
-            String packageNameForDeepLink = findAllowedPackageNameForDeepLink(url);
-            if (packageNameForDeepLink != null && isPackageNameInstalledAndEnabled(context, packageNameForDeepLink)) {
-              return Intents.createForOpeningUrl(url).setPackage(packageNameForDeepLink);
-            }
+              String packageNameForDeepLink = findAllowedPackageNameForDeepLink(url);
+              if (packageNameForDeepLink != null) {
+                  if (isPackageNameInstalledAndEnabled(context, packageNameForDeepLink))
+                      return Intents.createForOpeningUrl(url).setPackage(packageNameForDeepLink);
+                  else {
+                      /*
+                        third-party apps for only YouTube are supported at the moment.
+                        There can be more checks here if third-party apps for other services are supported too.
+                       */
+                      Intent youtubeIntent = Intents.createForYouTube(url);
+                      if (youtubeIntent != null && Intents.hasAppToHandleIntent(context, youtubeIntent))
+                          return youtubeIntent;
+                  }
+              }
```

Comment:
This code looks messy and is clearly going to get worse with more control flows. It also looks like a candidate for Polymorphism 101. 

I am thinking that the lines 129..142 can be replaced something like this:

```kotlin
// Get rid of the default web browser check. I think it's alright 
// to deeplink regardless of the selected browser.

if (isWebUrl) {
  val deeplink: Deeplink? = findAllowedDeeplink(url)

  if (deeplink == null) {
    defaultWebBrowser.intentForUrl(context, url, expandFromRect)

  } else {
    deeplink.intent()
  }
}

interface Deeplink {
  fun intent()
}

data class YouTubeDeeplink {
  override fun intent() {
    return if (isYouTubeInstalledAndEnabled()) {
      Intents.createForYouTube()
    } else {
	  // Dispatch to official app.
    }
  }
}

data class PlayStoreDeeplink {
  override fun intent() {
    // dispatch only to Play Store.
  }
}

and so on for other supported apps. 
```

What do you think? 

(Sorry for using Kotlin. I was low on time and haven't written Java in a long time)

Reply:
This does look like a much better approach. Which package should I place classes like `YouTubeDeeplink` etc in?

Reply:
Does `urlparser` sound like a good place?

Reply:
Yup